time-manager/System/TimeManager.hs

```diff
@@ -37,162 +39,158 @@ module System.TimeManager (
 
 import Control.Concurrent (mkWeakThreadId, myThreadId)
 import qualified Control.Exception as E
-import Control.Monad (void)
-import Control.Reaper
 import Data.IORef (IORef)
 import qualified Data.IORef as I
-import Data.Typeable (Typeable)
-import System.IO.Unsafe
 import System.Mem.Weak (deRefWeak)
 
+#if defined(mingw32_HOST_OS)
+import qualified GHC.Event.Windows as EV
+#else
+import qualified GHC.Event as EV
+#endif
+
 ----------------------------------------------------------------
 
 -- | A timeout manager
-data Manager = Manager (Reaper [Handle] Handle) | NoManager
+data Manager = Manager Int
 
--- | No manager.
 defaultManager :: Manager
-defaultManager = NoManager
+defaultManager = Manager 0
+
+isNoManager :: Manager -> Bool
+isNoManager (Manager 0) = True
+isNoManager _ = False
+
+----------------------------------------------------------------
 
 -- | An action to be performed on timeout.
 type TimeoutAction = IO ()
 
+----------------------------------------------------------------
+
 -- | A handle used by a timeout manager.
 data Handle = Handle
-    { handleManager :: Manager
-    , handleActionRef :: IORef TimeoutAction
-    , handleStateRef :: IORef State
+    { handleTimeout :: Int
+    , handleAction :: TimeoutAction
+    , handleKeyRef :: ~(IORef EV.TimeoutKey)
     }
 
-{-# NOINLINE emptyAction #-}
-emptyAction :: IORef TimeoutAction
-emptyAction = unsafePerformIO $ I.newIORef (return ())
-
-{-# NOINLINE emptyState #-}
-emptyState :: IORef State
-emptyState = unsafePerformIO $ I.newIORef Inactive
-
 emptyHandle :: Handle
 emptyHandle =
     Handle
-        { handleManager = NoManager
-        , handleActionRef = emptyAction
-        , handleStateRef = emptyState
+        { handleTimeout = 0
+        , handleAction = return ()
+        , handleKeyRef = undefined
         }
 
-data State
-    = Active -- Manager turns it to Inactive.
-    | Inactive -- Manager removes it with timeout action.
-    | Paused -- Manager does not change it.
+isEmptyHandle :: Handle -> Bool
+isEmptyHandle Handle{..} = handleTimeout == 0
 
 ----------------------------------------------------------------
 
 -- | Creating timeout manager which works every N microseconds
 --   where N is the first argument.
```

Comment:
```hs
-- | Creating timeout manager which works every N microseconds 
--   where N is the first argument.
--
--   Setting a timeout of 0 or less (or using 'defaultManager')
--   will act as if no timeout was set, so __NO__ 'TimeoutAction'
--   is run when using 'withHandle', and 'register' returns a
--   'Handle' that doesn't do anything.
```

Reply:
Wait... does the timeout manager run the `TimeoutAction` _EVERY_ N microseconds? Or only once?
I'd imagine it would only do it once, because that would be the timeout.

Or would something like the following be more accurate:
```
Create a timeout 'Manager' that takes the first argument as
microseconds. This is how long the timeout will take on first
invocation and which will be used to refresh the timer after
using 'tickle'.
```